rslib/backend.proto

```diff
@@ -756,6 +757,35 @@ message BuiltinSearchOrder {
     bool reverse = 2;
 }
 
+message FilterToSearchIn {
+    enum NamedFilter {
+        WHOLE_COLLECTION = 0;
+        CURRENT_DECK = 1;
+        ADDED_TODAY = 2;
+        STUDIED_TODAY = 3;
+        AGAIN_TODAY = 4;
+        NEW = 5;
+        LEARN = 6;
+        REVIEW = 7;
+        DUE = 8;
+        SUSPENDED = 9;
+        BURIED = 10;
+        RED_FLAG = 11;
+        ORANGE_FLAG = 12;
+        GREEN_FLAG = 13;
+        BLUE_FLAG = 14;
+        NO_FLAG = 15;
+        ANY_FLAG = 16;
+    }
+    oneof filter {
+        NamedFilter name = 1;
+        string tag = 2;
+        string deck = 3;
+        string note = 4;
+        uint32 template = 5;
+    }
```

Comment:
It took me a while to figure out how _filterFunc() was using this - the use of kwargs obscures what is valid input, and prevents us from knowing when we've passed in some incorrect value. Instead of extending setFilter() to expect kwargs, what about defining various helper functions that return a string, and using its existing args interface which just takes a list of strings? eg~

```
    def _named_filter(self, filter: NamedFilter) -> str:
        return self.col.backend.filter_to_search(FilterToSearchIn(name=filter))

    def _deck_filter(self, deck: str) -> str:
        return self.col.backend.filter_to_search(FilterToSearchIn(deck=deck))

    def _commonFilters(self):
        return self._simpleFilters(
            (
                (tr(TR.BROWSING_WHOLE_COLLECTION), self._named_filter(NamedFilter.WHOLE_COLLECTION),
```

But to be honest, what I originally had in mind was just keeping the existing strings, and simply having a function in the backend like:

make_string_term(key="deck", val='deck_with_"quotes') → '"deck:deck_with_\"quotes'
make_string_term(key=None, val="is:new") → "is:new"
make_string_term(key="is", val="new") → "is:new"

The former approach buys us some extra static typing, but is more work to implement. Either would be fine :-)

Reply:
Hmm, with the exception of _named_filter(), perhaps we don't need the other helpers and could just call the code directly in the calling function, eg _deckFilters() would call the filter_to_search routine itself

Reply:
With the second approach, I don't see much improvement to the current situation, so I'll go with the first one.

Of course, you're right about the kwargs. I only wonder, wouldn't it make more sense to implement these functions directly in rsbackend.py? Or do you want to keep that module concise, whereas all hope is already lost for browser.py? 😅

Also, I think it's neat to have `_commonFilters()` etc. only pass a name like `NamedFilter.WHOLE_COLLECTION` to `_simpleFilters()`, so we only have to call `_named_filter()` once there. (Maybe we can even rename `_simpleFilters()` to `_namedFilters()` since it coincides so well with `NamedFilter`.)

Reply:
> Hmm, with the exception of _named_filter(), perhaps we don't need the other helpers and could just call the code directly in the calling function, eg _deckFilters() would call the filter_to_search routine itself

Since we seem to have to call `filter_to_search` with `pb.FilterToSearchIn`, we can better just use a wrapper like `backend._deck_filter(deck: str)` instead of importing `pb.FilterToSearchIn` through rsbackend and then calling `backend.filter_to_search(FilterToSearchIN(deck="..."))` in my opinion.

Reply:
Re the appropriate file, rsbackend.py (and _gen.py) expose the various backend methods largely as is, and it's the job of the rest of pylib to wrap them and provide a more Pythonic API where necessary. Code in the GUI calling methods on .backend directly is a bit of a code smell - ideally every function would have a wrapper in pylib that we'd call instead. I haven't bothered with that in cases where the code is not intended to be user facing - which is partly laziness, and partly because Python doesn't let us enforce public/private methods - so even if we went to the trouble of wrapping backend methods that are not intended for public consumption, we can't stop users from ignoring the methods and calling the backend directly, making it feel a bit pointless.

Reply:
Thank you for taking the time to elaborate. It's good to know about this consideration.